naptime-graphql/src/main/scala/org/coursera/naptime/ari/graphql/GraphqlSchemaProvider.scala

```diff
@@ -0,0 +1,105 @@
+ * that they change relatively slowly. As a result, we do not bother taking a lock between querying
+ * the types and resourceSchemas collections. In case there is a change in collections between, we
+ * retry a fixed number of times, before giving up and operating with an inconsistent snapshot.
+ *
+ * Note: we rely on object identity to ensure an efficient set comparison. (This is a reasonably good
+ * approach, because we assume immutable collections. Therefore we know we will never skip re-computing
+ * when we should. If the schema provider returns sets that are not object identities, but are equal, then
+ * we unnecessarily re-compute. We feel this is reasonable, because if the SchemaProvider is not re-using
+ * the same sets, then that is generating unnecessary garbage, and that should be fixed.)
+ *
+ * @param schemaProvider A schema provider.
+ */
+class DefaultGraphqlSchemaProvider @Inject() (schemaProvider: SchemaProvider)
+  extends GraphqlSchemaProvider with StrictLogging {
+  private[this] var resourceSchemas: Set[Resource] = Set.empty
+  private[this] var typesSet: Set[DataSchema] = Set.empty
+  private[this] var cachedSchema: Schema[SangriaGraphQlContext, Any] = DefaultGraphqlSchemaProvider.EMPTY_SCHEMA
+
+  private[this] def consistentSnapshot(numRetries: Int = 2): (Set[Resource], Set[DataSchema]) = {
+    if (numRetries == 0) {
+      logger.warn(s"Could not compute a consistent snapshot of the schema provider.")
+      (schemaProvider.resourceSchemas, schemaProvider.types)
+    } else {
+      val resourceSchemas = schemaProvider.resourceSchemas
+      val types = schemaProvider.types
+      if ((schemaProvider.resourceSchemas eq resourceSchemas) && (schemaProvider.types eq types)) {
+        (resourceSchemas, types)
+      } else {
+        consistentSnapshot(numRetries - 1)
+      }
+    }
+  }
+
+  private[this] def recomputeSchema(): Unit = {
+    val (resourceSchemas, typesSet) = consistentSnapshot()
+    val typesMap = typesSet.collect {
+      case record: RecordDataSchema => record.getFullName -> record
+    }.toMap
+
+    val resources = resourceSchemas
+    val types = resources.flatMap { resource =>
+      typesMap.get(resource.mergedType).map(resource.mergedType -> _).orElse {
+        logger.warn(s"Did not find merged type `${resource.mergedType}` for resource " +
+          s"${resource.name}.v${resource.version.getOrElse(0L)}")
+        None
+      }
+    }.toMap
+    try {
+      val builder = new SangriaGraphQlSchemaBuilder(resources, types)
+      val graphQlSchema =
+        builder.generateSchema().asInstanceOf[Schema[SangriaGraphQlContext, Any]]
+      this.resourceSchemas = resourceSchemas
+      this.typesSet = typesSet
+      this.cachedSchema = graphQlSchema
+    } catch {
+      case NonFatal(e) =>
+        logger.error(s"Could not build schema.", e)
+        this.resourceSchemas = resourceSchemas
+        this.typesSet = typesSet
+        this.cachedSchema = DefaultGraphqlSchemaProvider.EMPTY_SCHEMA
```

Comment:
on a failure, do we want to fall back to the empty schema, or instead just keep the existing schema?